src/ragas/prompt/few_shot_pydantic_prompt.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+
+import typing as t
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+
+import numpy as np
+from pydantic import BaseModel
+
+from ragas.embeddings.base import embedding_factory
+from ragas.llms.base import BaseRagasLLM
+from ragas.prompt.pydantic_prompt import PydanticPrompt
+
+if t.TYPE_CHECKING:
+    from langchain_core.callbacks import Callbacks
+
+    from ragas.embeddings.base import BaseRagasEmbeddings
+    from ragas.llms.base import BaseRagasLLM
+
+# type variables for input and output models
+InputModel = t.TypeVar("InputModel", bound=BaseModel)
+OutputModel = t.TypeVar("OutputModel", bound=BaseModel)
+
+
+class ExampleStore(ABC):
+    @abstractmethod
+    def get_examples(
+        self, data: BaseModel, top_k: int = 5
+    ) -> t.Sequence[t.Tuple[BaseModel, BaseModel]]:
+        pass
+
+    @abstractmethod
+    def add_example(self, input: BaseModel, output: BaseModel):
+        pass
+
+
+@dataclass
+class InMemoryExampleStore(ExampleStore):
+    embeddings: BaseRagasEmbeddings = field(default_factory=embedding_factory)
+    _examples_list: t.List[t.Tuple[BaseModel, BaseModel]] = field(
+        default_factory=list, repr=False
+    )
+    _embeddings_of_examples: t.List[t.List[float]] = field(
+        default_factory=list, repr=False
+    )
+
+    def add_example(self, input: BaseModel, output: BaseModel):
+        # get json string for input
+        input_json = input.model_dump_json()
+        self._embeddings_of_examples.append(self.embeddings.embed_query(input_json))
```

Comment:
In my experiments embedding user_input only and whole input (containing user_input, response,etc) gave almost identical results. But I think logically it makes more sense to only embed user_input. Although this has to be revisited again later. Let us leave it like this for now.

Reply:
but what if there is no user_input?

Reply:
It's something we have to figure out, ie what's the most accurate way to embed few shot examples. Let's leave it like this now.